controllers/authconfigs_reconciler.go

```diff
@@ -0,0 +1,477 @@
+		}
+
+		var unauthorized *authorinov1beta2.DenyWithSpec
+		if response.Unauthorized != nil {
+			unauthorized = &response.Unauthorized.DenyWithSpec
+		}
+
+		authConfig.Spec.Response = &authorinov1beta2.ResponseSpec{
+			Unauthenticated: unauthenticated,
+			Unauthorized:    unauthorized,
+			Success: authorinov1beta2.WrappedSuccessResponseSpec{
+				Headers: authorinoSpecsFromConfigs(response.Success.Headers, func(config kuadrantv1beta3.MergeableHeaderSuccessResponseSpec) authorinov1beta2.HeaderSuccessResponseSpec {
+					return authorinov1beta2.HeaderSuccessResponseSpec{SuccessResponseSpec: config.SuccessResponseSpec}
+				}),
+				DynamicMetadata: authorinoSpecsFromConfigs(response.Success.DynamicMetadata, func(config kuadrantv1beta3.MergeableSuccessResponseSpec) authorinov1beta2.SuccessResponseSpec {
+					return config.SuccessResponseSpec
+				}),
+			},
+		}
+	}
+
+	// callbacks
+	if callbacks := authScheme.Callbacks; callbacks != nil {
+		authConfig.Spec.Callbacks = lo.MapValues(callbacks, func(v kuadrantv1beta3.MergeableCallbackSpec, _ string) authorinov1beta2.CallbackSpec {
+			return v.CallbackSpec
+		})
+	}
+
+	return authConfig
+}
+
+func authorinoSpecsFromConfigs[T, U any](configs map[string]U, extractAuthorinoSpec func(U) T) map[string]T {
+	specs := make(map[string]T, len(configs))
+	for name, config := range configs {
+		authorinoConfig := extractAuthorinoSpec(config)
+		specs[name] = authorinoConfig
+	}
+
+	if len(specs) == 0 {
+		return nil
+	}
+
+	return specs
+}
+
+func equalAuthConfigs(existing, desired *authorinov1beta2.AuthConfig) bool {
+	// httprouterule back ref annotation
+	existingAnnotations := existing.GetAnnotations()
+	desiredAnnotations := desired.GetAnnotations()
+	if existingAnnotations == nil || desiredAnnotations == nil || existingAnnotations[kuadrantv1beta1.AuthConfigHTTPRouteRuleAnnotation] != desiredAnnotations[kuadrantv1beta1.AuthConfigHTTPRouteRuleAnnotation] {
+		return false
+	}
+
+	// labels
+	existingLabels := existing.GetLabels()
+	desiredLabels := desired.GetLabels()
+	if len(existingLabels) != len(desiredLabels) || !labels.Set(existingLabels).AsSelector().Matches(labels.Set(desiredLabels)) {
+		return false
+	}
+
+	// spec
+	return reflect.DeepEqual(existing.Spec, desired.Spec)
+}
+
+// TODO(guicassolato): remove these functions below if we decide not to build conditions from the HTTPRouteRule + hostnames
+
+// authorinoConditionsFromHTTPRouteRule builds a list of Authorino conditions from a HTTPRouteRule and a list of hostnames
+// * Each combination of HTTPRouteMatch and hostname yields one condition.
+// * Rules that specify no explicit HTTPRouteMatch are assumed to match all requests (i.e. implicit catch-all rule.)
+// * Empty list of hostnames yields a condition without a hostname pattern expression.
+func authorinoConditionsFromHTTPRouteRule(rule gatewayapiv1.HTTPRouteRule, hostnames []gatewayapiv1.Hostname) []authorinov1beta2.PatternExpressionOrRef {
```

Comment:
Now that the lookup key to select the AuthConfig will be tightly associated with the route (i.e hostname + all other HTTP attributes), I don't see much value in adding route conditions at the top-level of the AuthConfig that repeat the ones already in the wasm config and therefore that are evaluated in the wasm module before sending (or not sending) the request to Authorino.

If anything, maybe only to cover for cases where conditions built ultimately from HTTPRouteMatches and hostnames have not been implemented yet. E.g.: https://github.com/Kuadrant/kuadrant-operator/blob/d90b34389e8dbe2719ae03897bc6a4dee720a1e2/pkg/wasm/utils.go#L143-L145

But I'd rather have all route conditions in the wasm config only (which would include implementing the ones above for query params.)

WDYT?

cc @eguzki @alexsnaps

Reply:
I guess you're talking about [`HTTPQueryParamMatch`](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1.HTTPQueryParamMatch), right? So we'd ideally provide some means to handle these from [`request.query`](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/advanced/attributes#request-attributes) indeed. We could provide some `Map<String, String>` view into that, doing the url decoding et al... while leaving the attribute untouched. 

But yes, I agree that ideally the whole "route rule matching" logic would ideally be entirely expressed as top-level predicates within the wasm config and be evaluated there so to avoid all unnecessary traffic (and have a single point of decision making wrt what, if any, `Policy`ies apply and how to dispatch to external services accordingly.

Reply:
Yes, `HTTPQueryParamMatch`.

Until we get CEL in, I think I'm gonna keep building redundant AuthConfig top-level conditions from the route matches then. At least this case of the query params will be taken care of in the meantime (if only for auth).

But, like you, I too hope we can deal with it in the wasm module (preferably in CEL, but whatever way works for me.)

Reply:
See https://github.com/Kuadrant/wasm-shim/issues/117 ... I hope someone @eguzki @didierofrivia can explore the space. Otherwise I'll work on it as soon as I have ... in the future at some point :P

Reply:
After speaking again with @alexsnaps, I realise what I said before won't work. This is because the wasm-shim wouldn't be able to distinguish between two route rule matches with and without query params, thus eventually triggering the wrong AuthConfig for some requests. This is a no-go.

I'm going to drop these unused funcs and let's get CEl in ASAP then.

Thanks @alexsnaps!